### PR TITLE
Disallow expressions of type `void` to be used in truthiness checks

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Fixes an offhand comment from #17892

This found a bug (!) in `session.ts` - `reloadFromFile` returned `void` so we were never hitting this line. I added return statements to `reloadFromFile` based on what I *think* it should do but it's not immediately apparent.
```ts
        private reload(args: protocol.ReloadRequestArgs, reqSeq: number) {
            const file = toNormalizedPath(args.file);
            const tempFileName = args.tmpfile === undefined ? undefined : toNormalizedPath(args.tmpfile);
            const info = this.projectService.getScriptInfoForNormalizedPath(file);
            if (info) {
                this.changeSeq++;
                // make sure no changes happen before this one is finished
                if (info.reloadFromFile(tempFileName)) {
                    this.doOutput(/*info*/ undefined, CommandNames.Reload, reqSeq, /*success*/ true);
/*---------         !!!!!!!!!!!!! unreachable! */
                }
            }
        }
```

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@typescript-bot test this

**Comment by RyanCavanaugh** [comment]
@weswigham is it possible to view the RWC diffs?

**Comment by weswigham** [comment]
Not easily (or without doubling the build time). I think I could probably surface them as an extra vsts pr without too much consternation - I'll look into it. Although we should also strive for usually having a clean RWC run, too. ❤️

**Comment by RyanCavanaugh** [comment]
New RWC failures:

https://github.com/firebase/firebase-js-sdk/blob/master/packages/database/src/core/util/Tree.ts#L156 - this function accepts an `(arg: T) => void` callback and tests its result - basically a type error but not a real bug per se in this file (may be an error in other places though)

VS Code has a similar issue: https://github.com/Microsoft/vscode/blob/master/src/vs/base/parts/tree/browser/treeDefaults.ts#L268

**Comment by weswigham** [comment]
@typescript-bot test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -24527,17 +24527,25 @@ namespace ts {
...
             checkGrammarStatementInAmbientContext(node);
 
-            checkExpression(node.expression);
+            checkTruthinessExpression(node.expression);
             checkSourceElement(node.statement);
         }
 
+        function checkTruthinessExpression(node: Expression) {
+            const type = checkExpression(node);
+            if (type === voidType) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: A better way to write this is `if (type.flags & TypeFlags.Void)` as we may at some point introduce other type instances that represent a `void` type.

**Comment by weswigham** [comment]
@typescript-bot test this

**Comment by RyanCavanaugh** [comment]
`lodash` RWC turned up one "not *obviously* broken" example (simplified):
```js
const noop = () => { };
const blah = () => 5;
var fn = Math.random() > 0.5 ? noop : blah;
if (fn()) {

}
```
Due to subtype reduction, `fn` is `() => void` and we issue an error.

RWC also found some code in VS Code (since fixed) where they were writing
```ts
assert.ok(fs.exists("foo"));
```
where `fs.exists` is `(filename: string, callback: (r: boolean) => void) => void`, so very much a real bug.

**Comment by ahejlsberg** [comment]
@RyanCavanaugh I agree the lodash example isn't *obviously* broken, but it definitely is suspicious given that a function returning anything can be assigned to a function returning `void`. So, with that in mind, I think it is a reasonable catch.

**Comment by chharvey** [comment]
I don't think this should have been done. A lot of times people use `void` in a logical OR operator to simplify code:

before:
```
let timesTwo = (x: number) => {
  console.log(`original: ${x}`);
  return x * 2;
}
```
after:
```
let timesTwo = (x: number) => console.log(`original: ${x}`) || x * 2;
```

because `console.log()` is of type `void`, TypeScript v3.1.1 throws this error:
> error TS1345: An expression of type 'void' cannot be tested for truthiness

**Comment by RyanCavanaugh** [comment]
Let's take all discussion to #26262

**Pull request merged**